scripts/xtensa-build-all.sh

```diff
@@ -113,171 +90,148 @@ set -e
 OLDPATH=$PATH
 WORKDIR="$pwd"
 
-# build platform
-for j in ${PLATFORMS[@]}
+# build platforms
+for platform in ${PLATFORMS[@]}
 do
 	HAVE_ROM='no'
 	DEFCONFIG_PATCH=''
-	if [ $j == "byt" ]
-	then
-		PLATFORM="baytrail"
-		ARCH="xtensa"
-		XTENSA_CORE="Intel_HiFiEP"
-		ROOT="$pwd/../xtensa-root/xtensa-byt-elf"
-		HOST="xtensa-byt-elf"
-		XTENSA_TOOLS_VERSION="RD-2012.5-linux"
-	fi
-	if [ $j == "cht" ]
-	then
-		PLATFORM="cherrytrail"
-		ARCH="xtensa"
-		XTENSA_CORE="CHT_audio_hifiep"
-		ROOT="$pwd/../xtensa-root/xtensa-byt-elf"
-		HOST="xtensa-byt-elf"
-		XTENSA_TOOLS_VERSION="RD-2012.5-linux"
-	fi
-	if [ $j == "bdw" ]
-	then
-		PLATFORM="broadwell"
-		ARCH="xtensa"
-		XTENSA_CORE="LX4_langwell_audio_17_8"
-		ROOT="$pwd/../xtensa-root/xtensa-hsw-elf"
-		HOST="xtensa-hsw-elf"
-		XTENSA_TOOLS_VERSION="RG-2017.8-linux"
-	fi
-	if [ $j == "hsw" ]
-	then
-		PLATFORM="haswell"
-		ARCH="xtensa"
-		XTENSA_CORE="LX4_langwell_audio_17_8"
-		ROOT="$pwd/../xtensa-root/xtensa-hsw-elf"
-		HOST="xtensa-hsw-elf"
-		XTENSA_TOOLS_VERSION="RG-2017.8-linux"
-	fi
-	if [ $j == "apl" ]
-	then
-		PLATFORM="apollolake"
-		ARCH="xtensa-smp"
-		XTENSA_CORE="X4H3I16w2D48w3a_2017_8"
 
-		# test APL compiler aliases and ignore set -e here
-		type xtensa-bxt-elf-gcc > /dev/null 2>&1 && true
-		if [ $? == 0 ]
-		then
-			HOST="xtensa-bxt-elf"
-		else
-			HOST="xtensa-apl-elf"
-		fi
-
-		ROOT="$pwd/../xtensa-root/$HOST"
-		XTENSA_TOOLS_VERSION="RG-2017.8-linux"
-		HAVE_ROM='yes'
-	fi
-	if [ $j == "skl" ]
-	then
-		PLATFORM="skylake"
-		ARCH="xtensa"
-		XTENSA_CORE="X4H3I16w2D48w3a_2017_8"
-
-		# test APL compiler aliases and ignore set -e here
-		type xtensa-bxt-elf-gcc > /dev/null 2>&1 && true
-		if [ $? == 0 ]
-		then
-			HOST="xtensa-bxt-elf"
-		else
-			HOST="xtensa-apl-elf"
-		fi
+	case $platform in
+		byt)
+			PLATFORM="baytrail"
+			ARCH="xtensa"
+			XTENSA_CORE="Intel_HiFiEP"
+			HOST="xtensa-byt-elf"
+			XTENSA_TOOLS_VERSION="RD-2012.5-linux"
+			;;
+		cht)
+			PLATFORM="cherrytrail"
+			ARCH="xtensa"
+			XTENSA_CORE="CHT_audio_hifiep"
+			HOST="xtensa-byt-elf"
+			XTENSA_TOOLS_VERSION="RD-2012.5-linux"
+			;;
+		bdw)
+			PLATFORM="broadwell"
+			ARCH="xtensa"
+			XTENSA_CORE="LX4_langwell_audio_17_8"
+			HOST="xtensa-hsw-elf"
+			XTENSA_TOOLS_VERSION="RG-2017.8-linux"
+			;;
+		hsw)
+			PLATFORM="haswell"
+			ARCH="xtensa"
+			XTENSA_CORE="LX4_langwell_audio_17_8"
+			HOST="xtensa-hsw-elf"
+			XTENSA_TOOLS_VERSION="RG-2017.8-linux"
+			;;
+		apl)
+			PLATFORM="apollolake"
+			ARCH="xtensa-smp"
+			XTENSA_CORE="X4H3I16w2D48w3a_2017_8"
+
+			# test APL compiler aliases and ignore set -e here
+			type xtensa-bxt-elf-gcc > /dev/null 2>&1 && true
```

Comment:
I'm pretty sure `A && true` is the same than `A`

Reply:
`A || true` ignores `set -e` but it does that by making `$?` always true which _is_ obviously not wanted here considering the next line checks `$?`

The previous code didn't work with `set -e`. It worked without `set -e` but only because `&& true` did nothing, ironic. Can you tig blame these lines and find what happened? May also help find a way to test them.

`if type -p xtensa-bxt-elf-gcc; then...` is the only way to be compatible with `set -e`.

Reply:
There's a longer `set -e` story in issue #2752

Reply:
@marc-hb i see that you have exposed many major issues here, I will be addressing and working on these over the weekend

Reply:
Thanks! Please don't get wrong: you probably do not want to fix every pre-existing issue that has been mentioned everywhere it can be found in the script. That would be a "Big Bang integration". As a rule of thumb you should fix pre-existing issues only in the code you're changing or moving, ideally in a separate commit before changing it or moving it.

Few hard and fast rules, lots of trade-offs.

Reply:
> I'm pretty sure `A && true` is the same than `A`

Bash doesn't exit when `A && true` would be used, but it will exit if `A` alone is used. Found this answer: https://stackoverflow.com/questions/37838273/evaluating-bash-exit-codes-behaviour. And as you said `|| true` would execute when failed and change the exit code. 



> The previous code didn't work with `set -e`. It worked without `set -e` but only because `&& true` did nothing, ironic. Can you tig blame these lines and find what happened? May also help find a way to test them.

Not sure what you mean by previous code, because this code works with `set -e`, it is already set in the script before this command is executed. Also, here's the commit introducing this https://github.com/thesofproject/sof/commit/cad86dc340447f8ae514233d50c921a21c989d6b

Also, why are we adding the `-p` option? On checking the help, it seems like it will ignore if it is an alias. Not sure if anyone would alias this, but should we be ignoring aliases?

Reply:
My bad about the `&& true`, thanks a lot for the explanation. I never can remember all the `set -e` exceptions, I wish this weren't one of them.

Please don't search and replace the whole file but in the bits you refactor please replace `&& true` with `|| true` anyway because `A || true`  reads as plain English, the `true` may actually run sometimes  and `|| true` would work even if that situation wasn't excluded by `set -e`, I mean `|| true` doesn't rely on some language subtlety like no-op `&& true` does.

The `-p` in `type -p` is not needed here, sorry for the noise, forget it. It's just the force of habit in: `foo=$(type -p bar)`. I forgot `-p` omits aliases and functions, thanks (that's one of the perks of reviewing code: learning goes both ways :-)  I never needed to test the existence of a function or (God forbid) an alias in a script.